Source/NSubstitute.Acceptance.Specs/CustomHandlersSpecs.cs

```diff
@@ -0,0 +1,236 @@
+                            return RouteAction.Continue();
+
+                        var args = call.GetArguments();
+                        args[0] = "refArg";
+
+                        return RouteAction.Return("xxx");
+                    }));
+
+            string refValue = "ref";
+            source.GetValueWithRef(ref refValue).Returns("42");
+
+            //act
+            refValue = "ref";
+            var result = source.GetValueWithRef(ref refValue);
+
+            //assert
+            Assert.That(result, Is.EqualTo("42"));
+        }
+
+        [Test]
+        public void Set_out_parameter_doesnt_affect_call_specification()
+        {
+            var source = Substitute.For<IValueSource>();
+            var router = SubstitutionContext.Current.GetCallRouterFor(source);
+
+            //Configure our handler to update "out" argument value
+            router.RegisterCustomCallHandler(state =>
+                new ActionHandler(
+                    call =>
+                    {
+                        if (call.GetMethodInfo().Name != nameof(IValueSource.GetValueWithOut))
+                            return RouteAction.Continue();
+
+                        var args = call.GetArguments();
+                        args[0] = "outArg";
+
+                        return RouteAction.Return("xxx");
+                    }));
+
+            string outArg;
+            source.GetValueWithOut(out outArg).Returns("42");
+
+            //act
+            string otherOutArg;
+            var result = source.GetValueWithOut(out otherOutArg);
+
+            //assert
+            Assert.That(result, Is.EqualTo("42"));
+        }
+
+        [Test]
+        public void Is_not_called_for_specifying_call()
+        {
+            //arrange
+            var source = Substitute.For<IValueSource>();
+            var router = SubstitutionContext.Current.GetCallRouterFor(source);
+
+            bool wasInvoked = false;
+            router.RegisterCustomCallHandler(state =>
+                    new ActionHandler(_ =>
+                    {
+                        wasInvoked = true;
+                        return RouteAction.Continue();
+                    }));
+
+            //act
+            source.MethodWithArgs(Arg.Any<string>(), Arg.Is("42"));
+
+            //assert
+            Assert.That(wasInvoked, Is.False);
+        }
+
+        [Test]
+        public void Auto_value_is_returned_if_skipped()
+        {
+            //arrange
+            var source = Substitute.For<IValueSource>();
+            var router = SubstitutionContext.Current.GetCallRouterFor(source);
+
+            router.RegisterCustomCallHandler(state =>
+                    new ActionHandler(_ => RouteAction.Continue()));
+
+            //act
+            var result = source.GetValue();
+
+            //assert
+            Assert.That(result, Is.Not.Null);
+        }
+
+        [Test]
+        public void First_added_handler_has_precedence()
```

Comment:
It might be not very intuitive if we reverse order. That is not a regular "Returns", rather that is a pipeline member. I'd expect that registered handlers are executed one by one, rather than in opposite order.

But if you feel that it might be confusing - NP, I'll change that.